fix(lsp): only disable inlay hints / diagnostics if no other clients are connected

catlee opened the pull request:
In https://github.com/neovim/neovim/pull/24469#issuecomment-1651992691, @stasjok observed that in the current implementation, if one LSP client disconnects from a buffer, then inlay_hints and diagnostics will be disabled.

In this PR I've tried to fix that by examining the other LSP clients attached to the current buffer in the `LspDetach` handler. NB the handler is called *before* the client detaches, so we expect the number of attached clients to be 1 if this is the last client in the process of disconnecting.

I also used a similar method in each feature's refresh logic to prevent making calls if no attached clients support the method. This can happen if e.g. you enable inlay hints on a buffer that doesn't have any clients connected that support it.

catlee committed:
fix(lsp): disable inlay hints/diagnostics when no other clients are connected

mfussenegger commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -11,6 +11,8 @@ local augroup = api.nvim_create_augroup('vim_lsp_diagnostic', {})
 
 local DEFAULT_CLIENT_ID = -1
 
+local PULL_DIAGNOSTIC_METHOD = 'textDocument/diagnostic'
```
There was just another PR merged that adds constants in lsp.protocol.Methods

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -2229,6 +2230,22 @@ function M._refresh(method, opts)
+---
+---@param bufnr integer buffer number
+---@param method string LSP method
+function M._get_clients_with_method(bufnr, method)
```
get_clients supports a `method` filter that can be used instead of this new method

MariaSolOs commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -11,6 +11,8 @@ local augroup = api.nvim_create_augroup('vim_lsp_diagnostic', {})
 
 local DEFAULT_CLIENT_ID = -1
 
+local PULL_DIAGNOSTIC_METHOD = protocol.Methods.textDocument_diagnostic
```
Yay for using the new string constants :D

MariaSolOs commented:
For my own learning: Does this take into account dynamic registration as well?

MariaSolOs commented:
> I'm not sure yet. Do you know of a good way to test dynamic registration? What servers currently support it?

I'm honestly not sure either of how to test it either lol. It seems like Solargraph supports dynamic capabilities based on [this discussion in Lazy](https://github.com/LazyVim/LazyVim/discussions/990).

justinmk commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -408,6 +408,17 @@ local function disable(bufnr)
+---@param opts table | nil Additional options to pass to util_refresh
+---@private
+local function _refresh(bufnr, opts)
+  if #vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic }) > 0 then
```
Can this check be added to `util._refresh` instead? Could be an option on `opts` if there is a case(?) where it makes sense to request things not supported by clients...

catlee replied:
yeah, good point. I'll move this check into `util._refresh`

justinmk commented on runtime/lua/vim/lsp/inlay_hint.lua:
```diff
@@ -131,6 +131,15 @@ local function disable(bufnr)
+---@param bufnr (integer) Buffer handle, or 0 for current
+---@private
+local function _refresh(bufnr)
+  if #vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_inlayHint }) > 0 then
```
same quesiton here

catlee commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -449,7 +460,12 @@ function M._enable(bufnr)
+        -- Only disable if there are no more clients with diagnostics attached
+        -- LspDetach is called *before* the client disconnects, so the number of clients connected
+        -- to this buffer that support diagnostics will be 1 if we're the last client
+        if #vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic }) == 1 then
```
Hmm, this doesn't always seem to happen.

I think I need to check `vim.lsp.get_clients` and filter out the client that is detaching, and then see if any remain.

justinmk commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -448,8 +457,18 @@ function M._enable(bufnr)
-      callback = function()
-        disable(bufnr)
+      callback = function(args)
+        local remaining_client_count = 0
+        for _, client in
+          pairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic }))
+        do
+          if client.id ~= args.data.client_id then
+            remaining_client_count = remaining_client_count + 1
+          end
+        end
+        if remaining_client_count == 0 then
+          disable(bufnr)
+        end
```
vim.iter may be worth using here (and in inlay_hint.lua).

```lua
local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic })
local remaining = vim.iter(clients):fold(0, function(acc, c)
  return (c.id == args.data.client_id) and i or i + 1
end)

if remaining == 0 then
  disable(bufnr)
end
```

stasjok replied:
> vim.iter may be worth using here (and in inlay_hint.lua).

Maybe like this?

```lua
local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_diagnostic })
if vim.iter(clients):any(function(client) return client.id ~= args.data.client_id end) then
  disable(bufnr)
end
```

catlee replied:
That's much nicer, thank you!

glepnir commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -2196,12 +2197,19 @@ end
   end
+
+  local force = opts.force or false
+  if (not force) and #vim.lsp.get_clients({ bufnr = bufnr, method = method }) == 0 then
```
```suggestion
  if not (opts.force or false) and #vim.lsp.get_clients({ bufnr = bufnr, method = method }) == 0 then
```

catlee commented:
When testing this out with some extra logging, I noticed that we're sending multiple requests for inlay_hints / diagnostics each time the document changes.

This happens because the `LspNotify` autocmd gets executed once per client notification. In the inlay_hints / diagnostics handlers, we end up calling `util._refresh`, which calls `vim.lsp.buf_request`, which sends a request to all connected clients that support the given method.

The result is that each client gets notified N times, where N is the number of clients connected that support the given method.

catlee commented:
I can think of a few solutions here:

1. Change when `LspNotify` is triggered. Instead of triggering it per client [here](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp.lua#L1584), trigger it after all clients have been notified. In the payload, `data.client_id` could be changed to `data.client_ids`.
2. Modify `vim.lsp.buf_request` to specify which client to send the request to
3. Introduce something like `vim.lsp.buf_request_single` to send a request to a single client
4. Implement debouncing logic inside `LspNofify` handlers. Ensure that we only call `vim.lsp.buf_request` once per X milliseconds. I'm not really a fan of this one.

Any other ideas, or thoughts on the above?

stasjok commented:
> * Modify `vim.lsp.buf_request` to specify which client to send the request to
> * Introduce something like `vim.lsp.buf_request_single` to send a request to a single client

There is a `request()` method on a client object for that. For example, I think you can do something like
```lua
local client = vim.lsp.get_client_by_id(data.client_id)
if client.supports_method(...) then
  client.request(_method_, params, _, bufnr)
end
```

catlee commented:
> > * Modify `vim.lsp.buf_request` to specify which client to send the request to
> > * Introduce something like `vim.lsp.buf_request_single` to send a request to a single client
> 
> There is a `request()` method on a client object for that. For example, I think you can do something like
> 
> ```lua
> local client = vim.lsp.get_client_by_id(data.client_id)
> if client.supports_method(_method_) then
>   client.request(_method_, params, _, bufnr)
> end
> ```

Yes, that may be the best solution. I was worried about needing to re-implement some of the request tracking / cancellation logic from `vim.lsp.buf_request`, but I don't think that's needed here.

catlee committed:
fix(lsp): refresh only the client associated with LspNotify event

This fixes the issue where the LspNotify handlers for inlay_hint /
diagnostics would end up refreshing all attached clients.

The handler would call util._refresh, which called
vim.lsp.buf_request, which calls the method on all attached clients.

Now util._refresh takes an optional client_id parameter, which is used
to specify a specific client to update.

This commit also fixes util._refresh's handling of the `only_visible`
flag. Previously if `only_visible` was false, two requests would be made
to the server: one for the visible region, and one for the entire file.

Co-authored-by: Stanislav Asunkin <1353637+stasjok@users.noreply.github.com>

Co-authored-by: Mathias Fußenegger <mfussenegger@users.noreply.github.com>

catlee commented:
I believe I've fixed the issue with the LSP clients being triggered multiple times per change. I adopted @stasjok's suggestion and call `client.request` directly.

I will be unable to address any questions or feedback for the next week or so.

stasjok commented:
Hey, I'm not a maintainer here, so you can ignore my comments, but here is my input:

1) Is `_refresh()` helper functions are adding much value instead of just `utils._refresh(ms.textDocument_inlayHint, { bufnr=bufnr })`?
2) In `utils._refresh()` you should be able to just:
```lua
local clients = vim.lsp.get_clients({ id = opts.client_id, bufnr = bufnr, method = method })
```

stasjok commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -2192,40 +2193,66 @@ end
-      }
-      vim.lsp.buf_request(bufnr, method, params)
+
+  local clients = {}
+  if opts.client_id then
+    local client = vim.lsp.get_client_by_id(opts.client_id)
+    if not (client and client.supports_method(method)) then
+      return
     end
+    vim.list_extend(clients, { client })
+  else
+    vim.list_extend(clients, vim.lsp.get_clients({ bufnr = bufnr, method = method }))
+  end
```
```suggestion
  local clients = vim.lsp.get_clients({ bufnr = bufnr, method = method, id = opts.client_id })
```

catlee replied:
Yeah, this is much nicer, thank you!

mfussenegger commented on runtime/lua/vim/lsp.lua:
```diff
@@ -2118,7 +2118,7 @@ api.nvim_create_autocmd('VimLeavePre', {
 ---@param method (string) LSP method name
 ---@param params table|nil Parameters to send to the server
----@param handler lsp-handler See |lsp-handler|
+---@param handler lsp-handler|nil See |lsp-handler|
```
```suggestion
---@param handler? lsp-handler See |lsp-handler|
```

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -2192,40 +2193,57 @@ end
 ---@private
 --- Request updated LSP information for a buffer.
 ---
+---@class RefreshOptions
```
```suggestion
---@class lsp.util.RefreshOptions
```

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -2192,40 +2193,57 @@ end
----@param opts (nil|table) Optional arguments
----  - bufnr (integer, default: 0): Buffer to refresh
----  - only_visible (boolean, default: false): Whether to only refresh for the visible regions of the buffer
+---@param opts? RefreshOptions Options table
```
```suggestion
---@param opts? lsp.util.RefreshOptions Options table
```

catlee commented:
Is there anything left to do before this can be merged in?
